worldedit-core/src/main/java/com/fastasyncworldedit/core/queue/implementation/ApplyTask.java

```diff
@@ -0,0 +1,241 @@
+
+    private static final int INITIAL_REGION_SHIFT = 5;
+    private static final int SHIFT_REDUCTION = 2;
+
+    private final CommonState<F> commonState;
+    private final ApplyTask<F> before;
+    private final int minChunkX;
+    private final int minChunkZ;
+    private final int maxChunkX;
+    private final int maxChunkZ;
+    // Note: shift == INITIAL_REGION_SHIFT means we are in the root node.
+    // compute() relies on that when triggering postProcess
+    private final int shift;
+
+    @Override
+    public void run() {
+        compute();
+    }
+
+    private record CommonState<F extends Filter>(
+            F originalFilter,
+            Region region,
+            ParallelQueueExtent parallelQueueExtent,
+            ConcurrentMap<Thread, ThreadState<F>> stateCache,
+            boolean full,
+            boolean[] faweExceptionReasonsUsed
+    ) {
+
+    }
+
+    private static final class ThreadState<F extends Filter> {
+
+        private final SingleThreadQueueExtent queue;
+        private final F filter;
+        private ChunkFilterBlock block;
+
+        private ThreadState(SingleThreadQueueExtent queue, F filter) {
+            this.queue = queue;
+            this.filter = filter;
+        }
+
+    }
+
+    ApplyTask(
+            final Region region,
+            final F filter,
+            final ParallelQueueExtent parallelQueueExtent,
+            final boolean full, final boolean[] faweExceptionReasonsUsed
+    ) {
+        this.commonState = new CommonState<>(
+                filter,
+                region.clone(), // clone only once, assuming the filter doesn't modify that clone
+                parallelQueueExtent,
+                new ConcurrentHashMap<>(),
+                full,
+                faweExceptionReasonsUsed
+        );
+        this.before = null;
+        final BlockVector3 minimumPoint = region.getMinimumPoint();
+        this.minChunkX = minimumPoint.x() >> 4;
+        this.minChunkZ = minimumPoint.z() >> 4;
+        final BlockVector3 maximumPoint = region.getMaximumPoint();
+        this.maxChunkX = maximumPoint.x() >> 4;
+        this.maxChunkZ = maximumPoint.z() >> 4;
+        this.shift = INITIAL_REGION_SHIFT;
+
+    }
+
+    private ApplyTask(
+            final CommonState<F> commonState,
+            final ApplyTask<F> before,
+            final int minChunkX,
+            final int maxChunkX,
+            final int minChunkZ,
+            final int maxChunkZ,
+            final int higherShift
+    ) {
+        this.commonState = commonState;
+        this.minChunkX = minChunkX;
+        this.maxChunkX = maxChunkX;
+        this.minChunkZ = minChunkZ;
+        this.maxChunkZ = maxChunkZ;
+        this.before = before;
+        this.shift = Math.max(0, higherShift - SHIFT_REDUCTION);
+    }
+
+    @Override
+    protected void compute() {
+        if (this.minChunkX != this.maxChunkX || this.minChunkZ != this.maxChunkZ) {
```

Comment:
`shift == 0` ? Or I haven't worked it out right

Reply:
` shift == 0` basically means "we're splitting this task into per-chunk-tasks now, so the current task could still cover multiple chunks